plugins/commands.js

```diff
@@ -0,0 +1,187 @@
+            auth: {
+                strategies: ['token'],
+                scope: ['user', 'build']
+            },
+            plugins: {
+                'hapi-swagger': {
+                    security: [{ token: [] }]
+                }
+            },
+            handler: (request, reply) => {
+                const namespace = request.params.namespace;
+                const name = request.params.name;
+                const version = request.params.version;
+
+                const id = `${namespace}-${name}-${version}`;
+
+                cache.get(id, (err, value) => {
+                    if (err) {
+                        return reply(err);
+                    }
+                    if (!value) {
+                        return reply(boom.notFound());
+                    }
+
+                    // @TODO put cache headers in here
+                    const response = reply(Buffer.from(value.c.data));
+
+                    response.headers = value.h;
+
+                    return response;
+                });
+            },
+            validate: {
+                params: {
+                    namespace: SCHEMA_COMMAND_NAMESPACE,
+                    name: SCHEMA_COMMAND_NAME,
+                    version: SCHEMA_COMMAND_VERSION
+                }
+            }
+        }
+    }, {
+        method: 'PUT',
+        path: '/commands/{namespace}/{name}/{version}',
+        config: {
+            description: 'Write command',
+            notes: 'Write a script or binary of specific command',
+            tags: ['api', 'commands'],
+            payload: {
+                maxBytes: parseInt(options.maxByteSize, 10) || DEFAULT_BYTES,
+                parse: false
+            },
+            auth: {
+                strategies: ['token'],
+                scope: ['build']
+            },
+            plugins: {
+                'hapi-swagger': {
+                    security: [{ token: [] }]
+                }
+            },
+            handler: (request, reply) => {
+                const pipelineId = request.auth.credentials.pipelineId;
+                const namespace = request.params.namespace;
+                const name = request.params.name;
+                const version = request.params.version;
+                const id = `${namespace}-${name}-${version}`;
+                const apihost = request.server.app.ecosystem.api;
+                const contents = {
+                    c: request.payload,
+                    h: {}
+                };
+                const size = Buffer.byteLength(request.payload);
+
+                // Store all x-* and content-type headers
+                Object.keys(request.headers).forEach((header) => {
+                    if (header.indexOf('x-') === 0 || header === 'content-type') {
+                        contents.h[header] = request.headers[header];
+                    }
+                });
+
+                return canPublishCommand(namespace, name, pipelineId, apihost)
```

Comment:
Is there ever a case when a user would only PUT a command to the store and not POST the command to the API? 
Can we replace the `canPublishCommand` with an API call to POST the command?

Reply:
For a new command, `sd-cmd` cli call both the API and Store when publishing a new command.
Other component metas, such as the template, are POSTed to API. To maintain consistency, I think that command meta should be POSTed to API, not to Store. Should the store be allowed to be POSTed command meta?

Reply:
I don't really see a reason why the Store shouldn't be able to POST the command meta. 
I was more concerned about having to make 2 API calls in total if the `sd-cmd` cli makes the POST call for command meta and the Store makes a GET to make sure permissions are okay and then stores the command. It also seems like we have to handle duplicate logic in two places (the API and the Store).

If we POST the command meta in the Store, it'll also handle permissions, and we won't need the `canPublishCommand` function anymore.

Not sure if @jithin1987 has an opinion about this?

Reply:
I would prefer API orchestrating the command creation and storing the command in store, with `sd-cmd` cli as a wrapper around API

Reply:
@jithin1987 So, should `sd-cmd` POST both the command meta and the command to API, then POST the command from API to Store?

Reply:
That's my recommendation, `sd-cmd` need not worry or orchestrate how command is stored let API handle all authentication and storage.

Reply:
I think that POST both the command meta and the command to API, then POST the command from API to Store is reasonable. Implementing so, the API can handle permissions, and the Store does only storing the command. Then we can remove the `canPublishCommand`.